pagetable creation/management

eim64 opened the pull request:
Paging is new set up to provide convenient way of creating pagetables, 
mapping and contstraining memory usage, which is a necessity for processes. 
Paging, as in swapping pages between disk and RAM will come later. 

Under the assumption that each process use their own pagetables, which
is described here:

https://forum.osdev.org/viewtopic.php?f=1&t=32996
https://en.wikipedia.org/wiki/Context_switch#Cost

Then setting up a new page table could look like this:

    PageEntry* new_p4 = create_pagetable();
    map_pages(new_p4, 0, program_size, program_start, PAGING_WRITEABLE | PAGING_USERSPACE | ...);
    map_pages(new_p4, data_start, data_end, program_start + data_start, PAGING_USERSPACE | ...);

... go through each section and set flags ...

then when the process is swapped, you can get or set the table used by the MMU

    set_cr3(new_p4);

when the process is killed, its pagetable can be freed

    free_pagetable(new_p4, 4);

antlilja commented on kernel/src/paging.c:
```diff
@@ -39,16 +80,44 @@ const PageFlags pagingmasks[5] = {
 };
 
+// returns if a pagetable entry can be expressed as a large page, rather than a table
+bool check_table_redundancy(PageEntry page_entry, uint64_t page_depth) {
```
I still don't think we should do this automatically. This is a very rare event, but can be a bit expensive to do every time. It isn't worth it in my opinion. It could also cause problems when multiple cores are using the same page table.  We could keep it as something to do manually.

eim64 replied:
This would be worthwhile for repeated usage of the same pagetable, but for the case of creating and discarding pagetables for each process it would indeed be unreasonable to include.

antlilja commented on kernel/src/paging.c:
```diff
@@ -18,17 +19,57 @@ void set_cr3(PageEntry* P4) {
-   should be available for new page tables.
-*/
-PageEntry* alloc_pagetable() { return page_zone_pointer += 512; }
+// leaks occur if too many tables are freed in total, this will count how many are leaked.
```
Is this because of your implementation or something unavoidable?

eim64 replied:
unavoidable without growing buffers.

antlilja replied:
What buffers?

eim64 replied:
g_free_tables_stack[FREED_TABLE_SIZE]

antlilja commented on kernel/include/paging.h:
```diff
@@ -12,29 +39,30 @@
+#define PAGETABLES_SIZE 0xFA000
+
+//this is where all pagetables are stored, we don't want to run out.
+extern uint64_t page_tables[PAGETABLES_SIZE];
```
these globals shouldn't be in the header because they don't need to be used outside of the source file

antlilja commented on kernel/src/paging.c:
```diff
@@ -18,17 +19,57 @@ void set_cr3(PageEntry* P4) {
-PageEntry page_zone_stack[PAGETABLE_SIZE] __attribute__((aligned(4096)));
-PageEntry* page_zone_pointer = &page_zone_stack[0];
+// stores pagetables
+PageEntry page_tables[PAGETABLES_SIZE] __attribute__((aligned(4096)));
```
Globals should have a g_ prefix

antlilja commented on kernel/include/paging.h:
```diff
@@ -55,3 +83,10 @@ void identity_map_pages(PageEntry* table, uint64_t virtaddr_low, uint64_t virtad
 PageEntry* fetch_page_entry(PageEntry* table, uint64_t virtaddr);
+
+// creates a new page table 
+PageEntry* create_pagetable();
```
Is this the first layer or what does this function do exactly? (Comment)

antlilja commented on kernel/src/paging.c:
```diff
@@ -4,6 +4,7 @@
 #define PAGETABLE_DEFAULT (PAGING_PRESENT | PAGING_WRITABLE)
 
 #define GET_TABLE_PTR(page_entry) (PageEntry*)(page_entry & PAGING_PTRMASK)
+#define GET_LAYER_PAGESIZE(depth) (1UL << (3 + 9 * depth))
```
depth should probably be inclosed in parens because this is a macro (otherwise order of operation could do weird things). This is also true for many of the other macros.

antlilja commented on kernel/src/paging.c:
```diff
@@ -148,7 +233,7 @@ void rec_setpage(PageEntry* table, uint64_t table_addr, uint8_t depth, MapParams
 
 void map_pages(PageEntry* table, uint64_t low, uint64_t high, uint64_t phys_low, PageFlags flags) {
     MapParams params = {
-        low & PAGING_PTRMASK, high & PAGING_PTRMASK, (phys_low - low) & PAGING_PTRMASK, flags};
```
Why can we just remove the subtraction here?

eim64 replied:
rec_setmap and MapParams was changed to avoid underflow and negative offsets.

the translated physical address now becomes
`const uint64_t physaddr = table_addr + params->physaddr_low - params->virtaddr_low;`

antlilja replied:
But why can't this underflow?

eim64 replied:
since all valid mappings will satisfy 
`table_addr + params->physaddr_low >= params->virtaddr_low`

antlilja replied:
Could you link me a source, I believe you I just to see where you found the information.

eim64 replied:
table_addr is in the bounds of [viraddrs_low, virtaddrs_high]. So it's natural that it's bigger or equal to virtaddrs_low.
I will however move it down to set_mapping, since it's the only place it's used.

antlilja replied:
What's the range of viraddr?

antlilja replied:
Because I'm not so sure that this can't underflow. You can map high virtual addresses to low physical addresses. Why can't this happen?

eim64 replied:
misspelled virtaddr previously.

table_addr is moved to virtaddrs_low at the start of the rec_setmap function, as described here
  
```c
// align table address with virutal address
// this is the same as finding the lower bound of a table
if (table_addr < params->virtaddr_low) {
    table_addr += params->virtaddr_low & pagingmasks[depth];
}
```
table_addr >= virtaddrs_low
therefore, 
table_addr + (positive number) - virtaddrs_low >= 0 

no underflow, QED.

antlilja replied:
That's what I wanted to see! It was really hard to see where the change that made sure of this occurred because there's only one big commit. In future try to create multiple commits with more contained changes, it'll make the life of the reviewer much easier and create less confusion. You can still create one PR but create multiple commits which describe what changes have been made and why.

antlilja commented on kernel/include/paging.h:
```diff
@@ -12,29 +39,26 @@
-//temporary solution before memory allocation is done
-extern uint64_t page_zone_stack[PAGETABLE_SIZE];
-extern uint64_t* page_zone_pointer;
+#define PAGETABLES_SIZE 0xFA000
```
This define does not need to be in the header

antlilja commented on kernel/include/paging.h:
```diff
@@ -2,6 +2,33 @@
+/*
+    This is currently set up to provide convenient way of creating pagetables, 
+    mapping and contstraining memory usage, which is a necessity for processes. 
+    Paging, as in swapping pages between disk and RAM will come later. 
```
I don't know if putting a page on disk while it's not being used is called "Paging". It's also not required, you put pages on disk when you run out of RAM or it isn't currently being used because loading a page from disk into ram on context switches is veeeeery slow. But maybe I'm misinterpreting this comment.

Commit:
pagetable allocation and freeing

outside code can now allocate and free entire pagetables.
when overwriting an entire pagetable, it will automatically be freed to
save memory.

Commit:
enclose macro args in parenthesees

Commit:
avoid intermediate underflow

it's not perticularly clear where a masked underflowed offset will end up.
it was thus rewritten to neither look underflowy.

antlilja committed:
Merge branch 'main' into pagetable-memory-managment
